pkg/hooks/manager.go

```diff
@@ -187,54 +187,69 @@ func (m *Manager) OnDefineDomain(domainSpec *virtwrapApi.DomainSpec, vmi *v1.Vir
 	if err != nil {
 		return "", fmt.Errorf("Failed to marshal domain spec: %v", domainSpec)
 	}
-	if callbacks, found := m.CallbacksPerHookPoint[hooksInfo.OnDefineDomainHookPointName]; found {
-		for _, callback := range callbacks {
-			if callback.Version == hooksV1alpha1.Version || callback.Version == hooksV1alpha2.Version {
-				vmiJSON, err := json.Marshal(vmi)
-				if err != nil {
-					return "", fmt.Errorf("Failed to marshal VMI spec: %v", vmi)
-				}
 
-				conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
-				if err != nil {
-					log.Log.Reason(err).Infof(dialSockErr, callback.SocketPath)
-					return "", err
-				}
-				defer conn.Close()
-
-				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
-				defer cancel()
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.OnDefineDomainHookPointName]
+	if !found {
+		return string(domainSpecXML), nil
+	}
 
-				switch callback.Version {
-				case hooksV1alpha1.Version:
-					client := hooksV1alpha1.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha1.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				case hooksV1alpha2.Version:
-					client := hooksV1alpha2.NewCallbacksClient(conn)
-					result, err := client.OnDefineDomain(ctx, &hooksV1alpha2.OnDefineDomainParams{
-						DomainXML: domainSpecXML,
-						Vmi:       vmiJSON,
-					})
-					if err != nil {
-						return "", err
-					}
-					domainSpecXML = result.GetDomainXML()
-				default:
-					panic("Should never happen, version compatibility check is done during Info call")
-				}
-			}
+	for _, callback := range callbacks {
+		var err error
+		domainSpecXML, err = m.onDefineDomainCallback(callback, domainSpecXML, vmi)
+		if err != nil {
+			return "", err
 		}
 	}
+
 	return string(domainSpecXML), nil
 }
 
+func (m *Manager) onDefineDomainCallback(callback *callBackClient, domainSpecXML []byte, vmi *v1.VirtualMachineInstance) ([]byte, error) {
+	if callback.Version != hooksV1alpha1.Version && callback.Version != hooksV1alpha2.Version {
+		panic("Should never happen, version compatibility check is done during Info call")
```

Comment:
Previously the code was not panicing on that condition. The iteration was just skipped.

Reply:
Ups, right. I was looking too much at `PreCloudInitIso`. Thanks for catching that.